scipy/integrate/_tanhsinh.py

```diff
@@ -0,0 +1,755 @@
+    # the level and `m` is an integer we get to choose. I choose
+    # m = _N_BASE_STEPS = `8` somewhat arbitrarily, but a rationale is that a
+    # power of 2 makes floating point arithmetic more predictable. It also
+    # results in a base step size close to `1`, which is what [1] uses (and I
+    # used here until I found [2] and these ideas settled).
+    h0 = tmax / _N_BASE_STEPS
+    return h0.astype(dtype)
+
+
+_N_BASE_STEPS = 8
+
+
+def _compute_pair(k, h0):
+    # Compute the abscissa-weight pairs for each level k. See [1] page 9.
+
+    # For now, we compute and store in 64-bit precision. If higher-precision
+    # data types become better supported, it would be good to compute these
+    # using the highest precision available. Or, once there is an Array API-
+    # compatible arbitrary precision array, we can compute at the required
+    # precision.
+
+    # "....each level k of abscissa-weight pairs uses h = 2 **-k"
+    # We adapt to floating point arithmetic using ideas of [2].
+    h = h0 / 2**k
+    max = _N_BASE_STEPS * 2**k
+
+    # For iterations after the first, "....the integrand function needs to be
+    # evaluated only at the odd-indexed abscissas at each level."
+    j = np.arange(max+1) if k == 0 else np.arange(1, max+1, 2)
+    jh = j * h
+
+    # "In this case... the weights wj = u1/cosh(u2)^2, where..."
+    pi_2 = np.pi / 2
+    u1 = pi_2*np.cosh(jh)
+    u2 = pi_2*np.sinh(jh)
+    # Denominators get big here. Overflow then underflow doesn't need warning.
+    # with np.errstate(under='ignore', over='ignore'):
+    wj = u1 / np.cosh(u2)**2
+    # "We actually store 1-xj = 1/(...)."
+    xjc = 1 / (np.exp(u2) * np.cosh(u2))  # complement of xj = np.tanh(u2)
+
+    # When level k == 0, the zeroth xj corresponds with xj = 0. To simplify
+    # code, the function will be evaluated there twice; each gets half weight.
+    wj[0] = wj[0] / 2 if k == 0 else wj[0]
+
+    return xjc, wj  # store at full precision
+
+
+def _pair_cache(k, h0):
+    # Cache the abscissa-weight pairs up to a specified level.
+    # Abscissae and weights of consecutive levels are concatenated.
+    # `index` records the indices that correspond with each level:
+    # `xjc[index[k]:index[k+1]` extracts the level `k` abscissae.
+    if h0 != _pair_cache.h0:
+        _pair_cache.xjc = np.empty(0)
+        _pair_cache.wj = np.empty(0)
+        _pair_cache.indices = [0]
+
+    xjcs = [_pair_cache.xjc]
+    wjs = [_pair_cache.wj]
+
+    for i in range(len(_pair_cache.indices)-1, k + 1):
+        xjc, wj = _compute_pair(i, h0)
+        xjcs.append(xjc)
+        wjs.append(wj)
+        _pair_cache.indices.append(_pair_cache.indices[-1] + len(xjc))
+
+    _pair_cache.xjc = np.concatenate(xjcs)
+    _pair_cache.wj = np.concatenate(wjs)
+    _pair_cache.h0 = h0
+
+_pair_cache.xjc = np.empty(0)
+_pair_cache.wj = np.empty(0)
+_pair_cache.indices = [0]
+_pair_cache.h0 = None
+
+
+def _get_pairs(k, h0, inclusive=False, dtype=np.float64):
```

Comment:
When log-integration is performed with complex integrand first, subsequent calls to regular integration can result in
> ComplexWarning: Casting complex values to real discards the imaginary part
  `return xjc[start:end].astype(dtype), wj[start:end].astype(dtype)`

Reply:
Basically. The root cause was that `h0` was complex, so I just added `real` to that.